File: Python/ch11/PocketMoneyRecur.py
def process(arr, index, rest):
    # 如果没有货币可选了
    if index == len(arr):
        # 恰好换完了，那么这也是一种方法，啥也不换
        # 否则，不是方法
        return 1 if rest == 0 else 0
    ways = 0
    zhang = 0
    while zhang * arr[index] <= rest:
        ways = ways + process(arr, index + 1, rest - (zhang * arr[index]))
        zhang = zhang + 1
    return ways


if __name__ == '__main__':
    a = [1, 2, 5, 10, 20, 50]
    total = 100
    print(process(a, 0, total))

File: Java/src/ch11/PocketMoneyRecur.java
package ch11;

public class PocketMoneyRecur {

    public static int process(int[] arr,int index,int rest){
        // 如果没有货币可选了
        if (index == arr.length){
            // 恰好换完了，那么这也是一种方法，啥也不换
            // 否则，不是方法
            return rest == 0 ? 1:0;
        }
        int ways = 0;
        for (int zhang = 0; zhang * arr[index] <= rest; zhang++) {
            ways = ways + process(arr,index+1,rest-(zhang*arr[index]));
        }
        return ways;
    }

    public static void main(String[] args) {
        int[] arr = {1,2,5,10,20,50};
        int total = 100;
        System.out.println(process(arr,0,total));
    }
}
